source/common/router/upstream_request.cc

```diff
@@ -370,12 +371,21 @@ void UpstreamRequest::maybeEndDecode(bool end_stream) {
   }
 }
 
-void UpstreamRequest::onUpstreamHostSelected(Upstream::HostDescriptionConstSharedPtr host,
-                                             bool pool_success) {
-  StreamInfo::UpstreamInfo& upstream_info = *streamInfo().upstreamInfo();
-  upstream_info.setUpstreamHost(host);
-  upstream_host_ = host;
-  parent_.onUpstreamHostSelected(host, pool_success);
+void UpstreamRequest::onUpstreamHostSelected(Upstream::HostDescriptionConstSharedPtr real_host,
+                                             bool success) {
+  // In most cases, the host description from host() of connection pool should be the same as
+  // the host description from the connection pool callback. Exception is the logical host
+  // which's addresses may be changed at runtime and the host description from the connection
+  // pool callback will be different from the host description from host() of connection pool.
+  if (real_host != nullptr && real_host != upstream_host_) {
```

Comment:
if we're latching host early but it may change, can this be a problem for folks latched to the old version?  Can you comment on why you want to latch early?

Reply:
Hah, may be can check if the upstream info has a valid upstream host when the upstream request is reset (remote or local), and if the upstream host is nullptr, we can get one from connection pool and set it to upstream info. WDYT?

Reply:
or is there a way of not snagging logical host?  or subscribing to updates so it's always kept up to date, or actually grabbing the outer abstraction instead of a handle which can change?  I think I'm OK with it changing if it must, but if we go that route we should probably also comment where the member is defined.

Reply:
For the logical host, when the connection is ready, the host will create a `RealHostDescrition` that wrap the logical host and the actual address that used by the connection.

This is a way to ensure the finally `HostDescrition` will always has an accurate address when logic host is used. Other types host needn't this because we assume it have a only one address.

However, with the support of the multiple-addresses of host, the correct way to get the accurate address of upstream host is the `Network::ConnectionInfoProvider` parameter of pool callbacks, for all types host.

So, I doubt the value of the `RealHostDescrition` now. Maybe we could remove the `RealHostDesription` directly and then the host from pool and the host from pool callbacks would be same. (And this would simplify the logic/complexity here, I taken some time to read the code, then found the difference between these hosts)

Reply:
cc @alyssawilk